# Call Center Cleanup

For this assignment, we will be working with call center data. You can start working on the assignment after the first lesson on Exploratory Data Analysis. Make sure to read the whole assignment before starting anything! As you code along in the Jupyter notebook, you are asked to make note of the results of your analysis. Do so by clicking on the results box and adding your notes beside each question.

## Business Issue and Understanding

You are working for a company that has two call centers: the North Call Center and the South Call Center. The company is looking to possibly hire five additional reps to enhance customer experience. Your task is to explore how efficient the current reps are in each branch to determine which branch would benefit from additional assistance.

### How the Call Center Works

Call center representatives are assigned queues. When calls are assigned to a queue, the call is assigned to the next person in line in the queue. After a call is assigned to a representative, the amount of time between assignment and the call starting is divided into busy minutes and not ready minutes. If the call is incoming and a customer is waiting on the phone for a rep, the time is split into three categories: busy minutes, not ready minutes, and incoming call wait time. Once the rep has the customer on the phone, there might be during call wait time, where the call is put on hold while the rep gets an answer for the customer.

### Notes about the Dataset

If you haven't worked in a call center before, these notes might help you throughout your analysis.

* The call purpose is tagged for each call.
* The time of the call is tagged in 1 hour blocks starting at 9:00 AM and ending at 5:00 PM.
* Calls are tagged as incoming or outgoing.
* Reps are assigned to queues. When the reps are working, they take calls in the order of their queue.
* A call that is dropped due to technical error or missed by the center because they have reached maximum capacity is a lost call.
* An abandoned call is when the customer hangs up because they have been waiting for too long.
* Busy Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are busy with other customers.
* Not Ready Minutes: the amount of time after a call comes in or needs to go out where the assigned rep is not available because they are not ready (for example, getting water).
* Incoming Wait Time - amount of time after assigned rep is available to take the call customer waits for representative to pick up a call. This is tracked in seconds.
* During Call Wait Time - amount of time during call that customer has to wait for representative

## Getting Started

You have two CSVs at your disposal, `NorthCallCenter.csv` and `SouthCallCenter.csv`. Import the appropriate libraries and create two dataframes, one called `north_df` and one called `south_df`.

In [48]:
# Import the appropriate libraries

import pandas as pd

# Create two new dataframes
north_df= pd.read_csv("NorthCallCenter.csv")
south_df= pd.read_csv ("SouthCallCenter.csv")

In [49]:
north_df.head()

Unnamed: 0 Branch     Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  North        Complaint    9:00 AM             Incoming     B   
1           1  North    Sales Support   11:00 AM             Incoming     A   
2           2  North    Sales Support   11:00 AM             Incoming     B   
3           3  North    Sales Support    4:00 PM             Incoming     B   
4           4  North  Product Support   12:00 PM             Incoming     B   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0  Brent   NO          0          1             9                  2   
1    Cam   NO          0          0            11                  1   
2   Todd   NO          0          0             9                  3   
3  Brent   NO          0          0            11                  2   
4  Brent   NO          0          0             8                  2   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      4      1  
1                 1.0                      3      1  
2                 1.0                      3      1  
3                 1.0                      4      1  
4                 1.0                      4      1

In [50]:
south_df.head()

Unnamed: 0 Branch   Call Purpose Time Block Incoming or Outgoing Queue  \
0           0  South  Sales Support   10:00 AM             Incoming     D   
1           1  South  Sales Support   10:00 AM             Incoming     C   
2           2  South  Sales Support   10:00 AM             Incoming     C   
3           3  South  Sales Support   10:00 AM             Incoming     C   
4           4  South  Sales Support   12:00 PM             Incoming     C   

  Rep ID Sale  Lost Call  Abandoned  Busy Minutes  Not Ready Minutes  \
0   Kate   NO          0          0             9                  1   
1   Eric   NO          0          0             8                  2   
2  Susan   NO          0          0            10                  2   
3  Alice   NO          0          0            12                  1   
4  Sandy   NO          0          0             8                  3   

   Incoming Wait Time  During Call Wait Time  Calls  
0                 1.0                      2      1  
1                 1.0                      4      1  
2                 1.0                      4      1  
3                 1.0                      3      1  
4                 1.0                      3      1

## Exploratory Data Analysis

Time to do some EDA! In the process of learning more about the two datasets, make sure you answer the following questions:

1. How many reps are in each branch?
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?
3. What is the number of calls taken for each time block?

In [51]:
#Reps in north branch
north_reps = north_df['Rep ID'].unique()
north_reps
len (north_reps)

9

In [52]:
#Reps in south branch 
south_reps = south_df['Rep ID'].unique()
south_reps
len (south_reps)

11

In [53]:

north_df['Busy Minutes'].mean()
#Avg busy minutes north


9.987755102040817

In [54]:
#Avg busy minutes south
south_df['Busy Minutes'].mean()

10.054140127388536

In [55]:
#avg not ready minutes north
north_df ['Not Ready Minutes'].mean()
 

1.9142857142857144

In [56]:
#Avg not ready minutes south 
south_df ['Not Ready Minutes'].mean()

1.910828025477707

In [57]:
#avg incoming wait time north
north_df ['Incoming Wait Time'].mean()

3.048780487804878

In [58]:
#avg incoming wait time south
south_df ['Incoming Wait Time'].mean()

3.0

In [59]:
#avg during call wait time north
north_df ['During Call Wait Time'].mean()

2.9673469387755103

In [60]:
#avg during call wait time south
south_df ['During Call Wait Time'].mean()

 

3.0764331210191083

In [61]:
#time_north = north_df['Time Block']
#time_north
#Time Block: amount of calls 
time_north = north_df.iloc[14:,3]
time_north

14     12:00 PM
15      1:00 PM
16      3:00 PM
17      2:00 PM
18     10:00 AM
         ...   
240     5:00 PM
241     5:00 PM
242     5:00 PM
243     5:00 PM
244     5:00 PM
Name: Time Block, Length: 231, dtype: object

In [62]:
#time_south = south_df ['Time Block']
#time_south

time_south = south_df.iloc[14:,3]
time_south

14      5:00 PM
15     11:00 AM
16     12:00 PM
17      1:00 PM
18      1:00 PM
         ...   
309     5:00 PM
310     5:00 PM
311     5:00 PM
312     5:00 PM
313     5:00 PM
Name: Time Block, Length: 300, dtype: object

### EDA Results

Make note of your results here!

1. How many reps are in each branch? North: 9, South: 11 
2. What is the average busy minutes, not ready minutes, incoming wait time, and during call wait time for each branch?
North Average busy minutes 9.987755102040817
South Average buy minutes 10.054140127388536

North not ready minutes 1.9142857142857144
South not ready minutes 1.910828025477707

North incoming wait time 3.048780487804878
South incoming wait time 3.0764331210191083

North during call wait time 2.9673469387755103
South during call wait time 3.0764331210191083

3. What is the number of calls taken for each time block?

North: 
14     12:00 PM
15      1:00 PM
16      3:00 PM
17      2:00 PM
18     10:00 AM
         ...   
240     5:00 PM
241     5:00 PM
242     5:00 PM
243     5:00 PM
244     5:00 PM

South: 
14      5:00 PM
15     11:00 AM
16     12:00 PM
17      1:00 PM
18      1:00 PM
         ...   
309     5:00 PM
310     5:00 PM
311     5:00 PM
312     5:00 PM
313     5:00 PM


In [63]:
#is question 3 answered properly?

## Cleaning Data

Now you need to clean up the datasets. When cleaning the datasets, you may find that there isn't dirty data to clean. That is okay! Some questions you need to answer about the data sets:

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them?
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?

In [64]:
north_df.info()
south_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 245 entries, 0 to 244
Data columns (total 15 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             245 non-null    int64  
 1   Branch                 245 non-null    object 
 2   Call Purpose           245 non-null    object 
 3   Time Block             245 non-null    object 
 4   Incoming or Outgoing   245 non-null    object 
 5   Queue                  245 non-null    object 
 6   Rep ID                 245 non-null    object 
 7   Sale                   245 non-null    object 
 8   Lost Call              245 non-null    int64  
 9   Abandoned              245 non-null    int64  
 10  Busy Minutes           245 non-null    int64  
 11  Not Ready Minutes      245 non-null    int64  
 12  Incoming Wait Time     82 non-null     float64
 13  During Call Wait Time  245 non-null    int64  
 14  Calls                  245 non-null    int64  
dtypes: flo

In [83]:
#Trying to answer question 1 for Data Cleaning Section. Want to isolate columns "Incoming or Outgoing [5]\ 
#and "Income Wait Time" [13] to compare "outgoing" and "null" respectively. 
#n_outgoing_calls = north_df.index = [5,13] did not work 
#Df- a collection of one or several series objects
#every column from the df is a series obj itself
null_wait_north = north_df [['Incoming Wait Time', 'Incoming or Outgoing']]
null_wait_north



Incoming Wait Time Incoming or Outgoing
0                   1.0             Incoming
1                   1.0             Incoming
2                   1.0             Incoming
3                   1.0             Incoming
4                   1.0             Incoming
..                  ...                  ...
240                 NaN             Outgoing
241                 NaN             Outgoing
242                 NaN             Outgoing
243                 NaN             Outgoing
244                 NaN             Outgoing

[245 rows x 2 columns]

In [ ]:
# Clean your data here
#numeric columns for north call center
n_numeric_cols= north_df.select_dtypes(include =['number']).columns
print (n_numeric_cols)

In [ ]:
#numeric columns for south call center 
s_numeric_cols = south_df.select_dtypes(include = ['number']).columns
print (s_numeric_cols)

In [ ]:
#North center non-numeric columns
n_non_numeric_cols = north_df.select_dtypes(exclude =['number']).columns
print (n_non_numeric_cols)

In [ ]:
#South center non-numeric columns
s_non_numeric_cols = south_df.select_dtypes(exclude = ['number']).columns
print (s_non_numeric_cols)

In [ ]:
#North center non-numeric columns null values
north_df[n_non_numeric_cols].info()
#no null values

In [ ]:
#South center non-numeric columns
south_df[s_non_numeric_cols].info()
#no null values

In [ ]:
north_df[n_numeric_cols].info()

In [ ]:
south_df[s_numeric_cols].info()

In [ ]:
num_missing = north_df.isna().sum()
num_missing

In [ ]:
num_missing = south_df.isna().sum()
num_missing

In [ ]:
#need to use a command to sum the 'outgoing' calls from column "Incoming or Outgoing" to answer the first question below

In [ ]:
#North Call Center without Incoming Wait Time, as this column was missing more than 50% of its values  
pct_missing = north_df.isna().mean()

pct_missing[pct_missing < .5]
north_df_less_missing_cols = north_df.loc[:, pct_missing <=.5].copy()
north_df_less_missing_cols

In [ ]:
#South Call Center without Incoming Wait Time, as this column was missing more than 50% of its values  
pct_missing = south_df.isna().mean()

pct_missing[pct_missing < .5]
south_df_less_missing_cols = south_df.loc[:, pct_missing <=.5].copy()
south_df_less_missing_cols

In [ ]:
#Unneccessary North Center Data, as determined by dropping rows with over 99.9% rows being same value
num_rows = len (north_df)

for col in north_df.columns:
    n_cnts = north_df[col].value_counts(dropna=False)
    n_top_pct = (n_cnts/num_rows).iloc[0]
    
    if n_top_pct > 0.999:
        print ('{0}: {1:.2f}%.format(col, n_top_pct*100)')
        print (n_cnts)
        print()
#Columns "Branch" and "Calls" were dropped as deemed unnnecessary due to having over 99.9% of the same values 
#in each row.

In [ ]:
#Unneccessary South Center Data, as determined by dropping rows with over 99.9% rows being same value
num_rows = len (south_df)

for col in south_df.columns:
    s_cnts = south_df[col].value_counts(dropna=False)
    s_top_pct = (s_cnts/num_rows).iloc[0]
    
    if s_top_pct > 0.999:
        print ('{0}: {1:.2f}%.format(col, s_top_pct*100)')
        print (s_cnts)
        print()
 #Columns "Branch" and "Calls" were dropped as deemed unnnecessary due to having over 99.9% of the same values 
#in each row.

### Cleaning Data Results

Make note of your results!

1. Incoming wait time is null in many places. Do all of those places correspond to outgoing calls?
2. In the North data set, there are two separate "YES" values for a sale. Why and how did you rectify the issue?
3. Are there duplicates in the two data sets? If there are, how did you handle them?
4. Is any of the data in the two data sets unnecessary? If yes, how did you handle it?
Unnecessary data — repetitive data, duplicates, and more
Columns "Branch" and "Calls" in both North Center and South Center df were deemed unnecessary due to having over 99.9% of rows containing same values.
Column "Incoming Wait Time was dropped from South Center and North Center df due to this column missing more than 50% of its values. 

## Data Manipulation

Before you begin answering any questions, combine the two datasets together to create a third dataframe called `df`. You can use this third dataframe to compare the two call centers to company-wide trends.

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each branch and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

In [ ]:
# Manipulate data here

### Data Manipulation Results

1. Group by Rep ID and sum the resulting structure. Sort by calls to determine which rep in each branch has the highest number of calls.
2. The average call abandonment rate is a KPI when it comes to determining call center efficiency. As you may recall, abandoned calls are calls where the customer hangs up due to long call times. What is the average call abandonment rate for each branch and the whole company? Do any of these fall out of the optimal range of 2-5%?
3. Service level is another KPI when it comes to measuring call center efficiency. Service level is the percentage of calls answered within a specific number of seconds. In the case of your employer, their ideal time frame is 2 seconds. What is the percentage of calls answered within 2 seconds for each company and the entire company?
4. For each branch and the entire company, what is the average speed of answer?

## Visualization

Create a visualization for each of the following questions. Some of the code to handle aggregating and storing data may be written for you. For each visualization, you choose the chart style that you feel suits the situation best. Make note of the chart style you chose and why.

1. What is the average abandonment rate per queue?
2. What is the service level and average speed of answer per each rep?
3. For each type of call purpose, how many calls are outgoing vs. incoming?

In [ ]:
# Create visualization 1 here
# The dictionary abandonment_rates has the data you need.

abandonment_rates = {}
queues = ["A", "B", "C", "D"]
queue_dict = df.groupby("Queue").agg("sum")
for i in range(4):
    abandonment_rates[queues[i]] = queue_dict["Abandoned"][i] / queue_dict["Calls"][i] 

In [ ]:
# Create visualization 2 here
# north_plt contains the data you need for the average speed of answer of each rep

north_plt = north.groupby("Rep ID")["Incoming Wait Time"].mean().to_frame().reset_index()

# Finding each Rep's Personal Service Level Percentage.  Basically, Calls within 2 secs / total calls

# Table 1: Total Incoming calls less than 2 seconds grouped by Rep
quick_calls = north[north["Incoming Wait Time"] <= 2.0]
quick_reps = quick_calls[["Rep ID", "Calls"]]
quick_stats = quick_reps.groupby(["Rep ID"]).sum()  # Final Table


# Table 2: Total Incoming Calls Only grouped by Rep
total_calls_in = north[north["Incoming or Outgoing"] == "Incoming"]
rep_calls = total_calls_in[["Rep ID", "Calls"]]     
total_stats = rep_calls.groupby(["Rep ID"]).sum() # Final Table  

#  Table 3: Service Level Percentage created via merge
service_level = pd.merge(quick_stats, total_stats, on="Rep ID")

# Create Percentage Column in Table 3
service_level["Service Level %"] = service_level["Calls_x"]/service_level["Calls_y"] * 100

In [ ]:
# Create visualization 3 here
# The three dictionaries, complaints, sales_support, and product_support, have the information you need

purpose_counts = purpose_group["Incoming or Outgoing"].value_counts()
complaints = purpose_counts["Complaint"].to_dict()
sales_support = purpose_counts["Sales Support"].to_dict()
product_support = purpose_counts["Product Support"].to_dict()

### Visualization Results

For each chart you created, explain why you chose the chart style you chose.

## Summarize Your Work

With what you know now about the two call centers and the entire company, answer the following questions. Note that while this is subjective, you should include relevant data to back up your opinion.

1. Using KPIs such as average abandonment rate, service level and average speed of answer, in your opinion, which one of the two branches is operating more efficiently? Why?
2. Based on the number of reps in each branch and how quickly the reps are working, in your opinion, which branch would benefit from the extra help?
3. Now that you have explored the datasets, is there any data or information that you wish you had in this analysis?

## Submit Your Work

After you have completed your work on the assignment, push your work to your Github repo. Navigate to the repo, copy the link to your assignment and paste the URL in the Canvas assignment's submission box. 

## Bonus Mission

Create a visualization that answers this question:

1. For each call purpose, how many calls (incoming and outgoing) take place in each time block?

In [ ]:
# Create your visualization here!
call_times = df[["Time Block", "Call Purpose", "Incoming or Outgoing", "Calls"]]

# Use groupby to plot based on time blocks:

# Use groupby and get_group to select which call purpose to plot:
